### PR TITLE
Cases advanced search

### LINKED ISSUE #4704: Advanced case search
**Is your feature request related to a problem in the current program to new available techology or software? Please describe and add links/citations if appropriate.**
It would be useful for institutes with many cases to be able to combine more than one criteria when filtering. E.g. both Tags, Status and Panel. 

**Describe the solution you'd like**
Either a separate search box or an "advanced" choice on the dropdown, and the ability to search with qualifiers directly, eg 
`tags: medical status: active panel: cardio`

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds a functionality or fixes a bug.
- Closes #4704

image

### Toggle advanced search:

image

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. Deploy on stage and try some advanced search

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

File: scout/adapter/mongo/case.py [hunk 1]
```diff
@@ -102,24 +103,20 @@ def _set_similar_phenotype_query(
...
 
         if len(similar_cases) == 0:  # No cases similar to given phenotype
             query["_id"] = {"$in": []}  # No result should be returned by query
             return
 
         similar_case_ids = []
-        order = []
         for i in similar_cases:
             similar_case_ids.append(i[0])
-            order.append(i[1])
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:116]: This is weird and I don't know how we were thinking when we wrote the function like this. 

Testing locally: `similar_cases` looks like this:

image

So `order` will be:  [1.0, 1.0, 1.0]

This value is returned by the function.

File: scout/adapter/mongo/case.py [hunk 2]
```diff
@@ -494,21 +487,20 @@ def _conditional_set_query_value(query, condition, set_key, set_value):
...
 
         if has_rna_data:
             cases_with_rna = self.rna_cases(collaborator or owner)
-            self._update_case_id_query(query, cases_with_rna)
+            self.update_case_id_query(query, cases_with_rna)
 
         if yield_query:
             return query
 
-        if order:
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:510]: Then order is used here, just to say if the array has values use the following projection:

'analysis_date': 1, 'assignees': 1, 'beacon': 1, 'case_id': 1, 'display_name': 1, 'individuals': 1, 'is_rerun': 1, 'is_research': 1, 'mme_submission': 1, 'owner': 1, 'panels': 1, 'status': 1, 'track': 1, 'vcf_files': 1}

I don't think the order thing adds anything by itself in here?

File: scout/adapter/mongo/case.py [hunk 3]
```diff
@@ -102,24 +103,20 @@ def _set_similar_phenotype_query(
                 hpo_terms.append(term.get("phenotype_id"))
             similar_cases = self.cases_by_phenotype(hpo_terms, institute_id, case_obj["_id"])
         else:  # similar HPO terms
-            hpo_terms = list(query_term.replace(" ", "").split(","))
+            hpo_terms = list(query_value.replace(" ", "").split(","))
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:106]: Just renamed he variable to be more clear

**Comment by northwestwitch** [comment]
@dnil unless you have started reviewing let's not include this PR in the next release I think

**Comment by dnil** [comment]
> @dnil unless you have started reviewing, let's not include this PR in the next release I think

Sounds like a plan! Will be really good, but may take a little more iterating. Query changes and all that.. 😅

File: scout/adapter/mongo/case.py (hunk 3 above)
Comment (dnil) [reply scout/adapter/mongo/case.py:106]: 💯

Comment (dnil) [reply scout/adapter/mongo/case.py:106]: Query term was more logical when the term was in there, and not in a separate field I guess.. 😁

File: scout/adapter/mongo/case.py [hunk 4]
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
...
-            self._set_case_name_query(query, query_term)
-
-        if query_field == "exact_pheno":
-            self._set_case_phenotype_query(query, query_term)
-
-        if query_field == "exact_dia":
-            self._set_diagnosis_query(query, query_term)
+    def update_case_query_or_options(self, query: dict, or_options: List[dict]):
+        """Populates the available options in the query $and field"""
+        if or_options and query.get("$or"):
```
Comment (dnil) [inline scout/adapter/mongo/case.py:216]: i
```suggestion
        if not or_options:
            return
       if query.get("$or"):
```

File: scout/adapter/mongo/case.py [hunk 5]
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
...
+            query["$or"] += or_options
+        elif or_options:
```
Comment (dnil) [inline scout/adapter/mongo/case.py:218]:
```suggestion
```

File: scout/adapter/mongo/case.py [hunk 6]
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
...
+                "user",
+            ]:
+                query_value = name_query.get(query_field)
+                if query_value not in ["", None]:
+                    set_case_item_query(
+                        query=query, query_field=query_field, query_value=query_value.strip()
+                    )
 
-    def _update_case_id_query(self, query, id_list):
+    def update_case_id_query(self, query, id_list):
```
Comment (dnil) [inline scout/adapter/mongo/case.py:296]: Huh, is this really supposed to be imported outside the module now? Or am I being archaic with PEP-8?
Screenshot 2024-08-28 at 08 12 57

File: scout/adapter/mongo/case.py [hunk 7]
```diff
@@ -201,94 +193,107 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
...
-                    {"phenotype_groups": {"$exists": False}},
-                ]
-        if query_field == "cohort":
-            query["cohorts"] = query_term
-
-        if query_term != "" and query_field in ["similar_case", "similar_pheno"]:
-            order = self._set_similar_phenotype_query(
-                query, query_field, query_term, owner or collaborator
-            )
+    def populate_case_query(
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/case.py:221]: Also it's not quite working? I tried filling in a couple of search fields: anything genes related gets a 
Screenshot 2024-08-28 at 08 27 39

File: scout/adapter/mongo/case.py [hunk 8]
```diff
@@ -521,7 +522,7 @@ def rna_cases(self, owner):
         Returns:
             list of case _ids
         """
-        EXISTS_NOT_NULL = {"$exists": True, "$nin": [None, ""]}
+        EXISTS_NOT_NULL = {EXISTS: True, "$nin": [None, ""]}
```
Comment (dnil) [inline scout/adapter/mongo/case.py:525]: I dunno, this looks like pleasing some silly code evaluation? It's not like mongo are going to change that function anytime soon?! And if so we could search-replace it just as easy. The rule is for different situations, like natural language to be translated, or more complex ones that could go wrong. "EXISTS_NOT_NULL" is good, and e.g. `EXISTS = {"$exists": True}` or whatnot would be ok in my book!

File: scout/server/blueprints/institutes/controllers.py [hunk 9]
```diff
@@ -10,13 +10,7 @@
...
 
 from scout.adapter.mongo.base import MongoAdapter
-from scout.constants import (
-    CASE_SEARCH_TERMS,
-    CASE_STATUSES,
-    DATE_DAY_FORMATTER,
-    ID_PROJECTION,
-    PHENOTYPE_GROUPS,
-)
+from scout.constants import CASE_STATUSES, DATE_DAY_FORMATTER, ID_PROJECTION, PHENOTYPE_GROUPS
```
Comment (dnil) [inline scout/server/blueprints/institutes/controllers.py:13]: 👍

File: scout/server/blueprints/institutes/controllers.py [hunk 10]
```diff
@@ -435,20 +429,14 @@ def cases(store, request, institute_id):
...
-                request.args["search_term"].strip(),
-            ]
-        )
-    data["name_query"] = name_query
+
+    name_query = request.form
     limit = int(request.args.get("search_limit")) if request.args.get("search_limit") else 100
-    data["form"] = populate_case_filter_form(request.args)
+
+    data["form"] = CaseFilterForm(request.form)
```
Comment (dnil) [inline scout/server/blueprints/institutes/controllers.py:439]: 👍

File: CHANGELOG.md [hunk 11]
```diff
@@ -22,6 +22,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 
 ## [4.87]
 ### Added
+- Advanced cases search to narrow down results using more than one search parameter
```
Comment (dnil) [inline CHANGELOG.md:25]: Careful with the changelog here! 😊

File: scout/adapter/mongo/case.py (hunk 6 above)
Comment (northwestwitch) [reply scout/adapter/mongo/case.py:319]: Mmm no, perhaps I intended to import it from some other place at one point, but I didn't do it. I'll revert!

File: scout/adapter/mongo/case.py (hunk 5 above)
Comment (northwestwitch) [reply scout/adapter/mongo/case.py]: I fixed it like this:

```
        if not or_options:
            return
        query.setdefault("$or", []).extend(or_options)
```

File: scout/adapter/mongo/case.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/adapter/mongo/case.py:220]: ChatGPT! 😆 I modified and removed the lambdas, but it wasn't so hard to follow I think. I mean, historically I'm the one no getting the lambda thing, so I reasoned that if I understood then everybody could. But sure, without lambdas is more readable!

File: scout/adapter/mongo/case.py (hunk 8 above)
Comment (northwestwitch) [reply scout/adapter/mongo/case.py:548]: Yes, of course it's pleasing that code checker, I would never think of such replacement myself. I like your suggestion but if you check in the code you'll find that the "exists" is used in slightly different situations, for instance:

- set_value={EXISTS: False}
- set_value={EXISTS: True, "$ne": []} # there are a bunch of these
- set_value={EXISTS: True, "$eq": True},
- EXISTS_NOT_NULL = {EXISTS: True, "$nin": [None, ""]}

That's why I decided to replace only "$exists" and not the whole dict. This way it is more easy to read and doesn't trigger the code checker. I'd like to leave it like this if it's not a problem for you

File: scout/adapter/mongo/case.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dnil) [reply scout/adapter/mongo/case.py:220]: I do like lambdas for functional stuff, like in sorting or for more map-reduce like operation, but mixed in for some functions in a large list, I think this is cleaner.

**Comment by northwestwitch** [comment]
I'm ready for another review with this one. And I guess we can include it in the next minor, so I
'm not going to merge in the upcoming patch!

File: tests/adapter/mongo/test_case_handling.py [hunk 12]
```diff
@@ -209,12 +248,14 @@ def test_get_research_case(real_adapter, case_obj, institute_obj):
     ## THEN assert that the case was inserted
     assert adapter.case_collection.find_one()
 
-    # THEN searching for reasearch cases should return one case
+    # THEN searching for research cases should return one case
```
Comment (dnil) [inline tests/adapter/mongo/test_case_handling.py:251]: 👌

**Review by dnil — COMMENTED** [review]
Definitely getting there! Works for me, but I have a couple of comments regarding the handling and search feeling during testing, and also some small comment to the code. It's still a little bit back and forth with local functions and "suggested private" class methods as well, but good enough methinks, and a less strenuous read than the initial AI code.
- It can be a bit annoying that the advanced query accordion folds back in after search if you wanted to test a few different things.
- This is compounded by values staying populated - as they should. It is too easy to get confused why cases don't show up and realise you at one point typed something in an "advanced" field previously and forgot about it. Perhaps force the advanced tab open if anything is filled in, or indicate somewhere next to the case name search field that advanced options are in play?
- A "clear search" option would be useful with this many fields, but could be added later.

File: scout/server/blueprints/institutes/templates/overview/cases.html [hunk 13]
```diff
@@ -22,49 +22,116 @@
...
 {% macro search_form() %}
-<form action="{{ url_for('overview.cases', institute_id=institute._id) }}" method="GET" accept-charset="utf-8">
-    <div class="row d-flex align-items-end">
-      <div class="col-3">
-        {{ form.search_type.label(class="control-label") }}
-        {{ form.search_type(class="form-control") }}
-      </div>
-      <div class="col-3">
-          <label class="control-label" for="search_term">Search cases</label>
-          <input class="form-control" id="search_term" name="search_term" value="{{form.search_term.data if form.search_term.data}}" placeholder="example:18201" type="text" pattern="[^\\\<\>\?\!\=\/]*" title="Characters \<>?!=/ are not allowed">
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/cases.html:33]: Ah right, good catch!, I'll re-add it!

File: scout/adapter/mongo/case.py (hunk 8 above)
Comment (northwestwitch) [reply scout/adapter/mongo/case.py:568]: I feel that if I replace the cases above with 5 different constants then it will be much less readable. If you don't like the EXISTS = "$exists" replacement I can simply revert to how the code was before and we ignore the code checker warning instead

File: scout/server/blueprints/institutes/templates/overview/cases.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/cases.html:33]: Fixed!

**Comment by northwestwitch** [comment]
> Definitely getting there! Works for me, but I have a couple of comments regarding the handling and search feeling during testing, and also some small comment to the code. It's still a little bit back and forth with local functions and "suggested private" class methods as well, but good enough methinks, and a less strenuous read than the initial AI code.
> 
> * It can be a bit annoying that the advanced query accordion folds back in after search if you wanted to test a few different things.
> * This is compounded by values staying populated - as they should. It is too easy to get confused why cases don't show up and realise you at one point typed something in an "advanced" field previously and forgot about it. Perhaps force the advanced tab open if anything is filled in, or indicate somewhere next to the case name search field that advanced options are in play?
> * A "clear search" option would be useful with this many fields, but could be added later.

Good suggestions, I fixed it as you said:

image

- filters stays open when you perform an advanced search
- At the bottom it shows the keys/values used in the search
- Added a clear search button

File: scout/adapter/mongo/case.py (hunk 2 above)
Comment (dnil) [reply scout/adapter/mongo/case.py:510]: Hm, I'm probably slow today, but at least I'm looking.. 😆 It returns them in the order given, without sorting, and the intention appears to be that they should appear in the same order as the "case_id: $in" operation gave them previously. Are you saying this doesn't work anyway, or that the returned similarity values are the same so it doesn't work? 😊

File: scout/adapter/mongo/case.py (hunk 8 above)
Comment (dnil) [reply scout/adapter/mongo/case.py:568]: Whatever! It's just a constant string regardless, and the only reason to change should be to make it more readable and maintainable. Since there is a "EXISTS_NOT_NULL" already, adding the others as similar full constants would make sense to me, but its no big deal.

File: scout/server/blueprints/institutes/templates/overview/cases.html [hunk 14]
```diff
@@ -27,7 +27,7 @@
     <div class="row">
         <div class="col-md-3 mb-3">
             {{ form.case.label(class="form-label") }}
-            {{ form.case(class="form-control", placeholder="example:18201") }}
+            {{ form.case(class="form-control", placeholder="example:18201", pattern="[^\\\<\>\?\!\=\/]*", title="Characters \<>?!=/ are not allowed") }}
```
Comment (dnil) [inline scout/server/blueprints/institutes/templates/overview/cases.html:40]: Hm, I wonder if we needed it for the other query fields as well?

Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/cases.html:40]: Fix, thanks. Better to be safe than sorry!

**Comment by northwestwitch** [comment]
Searching by similar case is actually broken, like you get an exception. I'll push a fix (not the fix to #4839, but a simple one)

**Comment by dnil** [comment]
Ahhh, I was about to say it works on stage, but just tested on prod. It no longer (if ever) works since the empty match can become `None` instead of `[]`. Good catch.

**Comment by northwestwitch** [comment]
> Searching by similar case is actually broken, like you get an exception. I'll push a fix (not the fix to #4839, but a simple one)

It happens when you look for cases similar to a case without phenotype. An existing bug I presume

**Comment by northwestwitch** [comment]
This one should be ready also @dnil

**Comment by dnil** [comment]
Hm, it still removes the old, arguably anyway partly broken order. Let's discuss that one - I would prefer to fix or remove it as part of #4882, but either fully or not here. I'm open to arguments though!

**Comment by northwestwitch** [comment]
> Hm, it still removes the old, arguably anyway partly broken order. Let's discuss that one - I would prefer to fix or remove it as part of #4839 , but either fully or not here. I'm open to arguments though!

I don't want to fix that order thing on this PR. It's already quite long. I'll attempt a fix in a new PR

File: scout/adapter/mongo/case.py [hunk 15]
```diff
@@ -108,13 +102,21 @@ def _set_similar_phenotype_query(
...
         else:  # similar HPO terms
-            hpo_terms = list(query_term.replace(" ", "").split(","))
+            hpo_terms = list(query_value.replace(" ", "").split(","))
             similar_cases = self.cases_by_phenotype(hpo_terms, institute_id, None)
 
-        query["_id"] = {"$in": [similar_case[0] for similar_case in similar_cases]}
+        similar_case_ids = [
+            similar_case[0] for similar_case in similar_cases
+        ]  # These are sorted by the most similar to the least
+        self._update_case_id_query(
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:111]: This advanced case search is quite delicate, because if you have other params that will affect the content of query["_id"]["$in"] you can mess up the order of the similar cases

File: scout/adapter/mongo/case.py [hunk 16]
```diff
@@ -197,86 +194,149 @@ def _set_diagnosis_query(self, query: Dict[str, Any], query_term: str):
...
+                "panel",
+                "status",
+                "tags",
+                "track",
+                "pheno_group",
+                "cohort",
+                "pinned",
+                "causative",
+                "user",
+                "similar_case",  # In order to be able to sort results by phenotype similarity, keep this at the bottom
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:331]: I have moved these 2 at the bottom so whatever happens before setting the query _id values, the order by similarity will be maintained (hopefully)

**Comment by northwestwitch** [comment]
I've included also the "search by case similarity" fix now. Tested with the combo search on cust003

{'cohort': 'hello', 'similar_case': 'holydassie', 'search_limit': 100}:

And it works, returning this case -> cout-stage.scilifelab.se/cust003/16062

**Review by dnil — APPROVED** [review]
Super, thank you, and especially for lifting the order fix part to the other PR! As before, this will be very useful, EXIST or not. 😁 Let's also make an issue for moving the dashboard search to this format, but no need to do in this PR. 😊

**Comment by northwestwitch** [comment]
> Let's also make an issue for moving the dashboard search to this format, but no need to do in this PR

Exactly, and if we do that the code will look clearer!
